### ISSUE TITLE
[Question] What is the idiomatic way to assert items count in lists with virtual rendering?

### ISSUE THREAD

**Issue by halexiev-hedgeserv** [microsoft/playwright#17042]
Hello,

We have numerous components (grid, list, etc.) that do not render all items at once but only the visible items in the viewport of the component. For example Angular's cdk-virtual-scroll-viewport does not have a way to turn of virtual rendering.

What would be Playwright's idiomatic way of asserting items count in such cases?

Thanks!

**Comment by pavelfeldman** [comment]
There is no items count in a virtual viewport as far as Playwright (or a user) is concerned. It can be 100 or a 1000, only the app's internals knows it, so you should query app internals using `page.evaluate()`.

**Comment by halexiev-hedgeserv** [comment]
Ok, but how would you use the evaluate method? Our approach would have been maybe to scroll a viewport at a time and count the items till we scroll to the end, but that seemed a bit flaky?

**Comment by shwarcu** [comment]
@halexiev-hedgeserv Hi, have you considered preparing version of your page/view without virtualization?
I am asking because what you might be trying to achieve is testing if external library/component is correctly working. At the same time you do not have any control over that external component (virtualized list).

I know this solution may appear to some people wrong, but I would propose to prepare some sort of feature switch, for example by adding a local storage entry such as `virtualization:off` which would make your app render components without virtualization.
It requires a bit of additional development to prepare non-virtualized views, however it will increase testability of your app by tools such as Playwright, Cyperss etc.
tl;dr of what I'm trying to suggest, is to avoid testing things that you do not have controll and also trying to remove such things from the list of things that affect your AUT. Virualization might be such thing that's worth to skip when testing automatically.

**Comment by halexiev-hedgeserv** [comment]
@shwarcu Hi, yes this is how we test the ag-grid component - it supports turning off the column virtualisation. This is not the case with Angular's cdk-virtual-scroll-viewport component though. We could not find a way to make it render all items it contains so that's why we were interested if there is any built-in support for such cases in Playwright. 

The suggested page.evaluate() method is a bit unfamiliar to us, so if somebody could elaborate on how to use it we could give it a try. As I mentioned above we tried with controlling the scrollbar and things were a bit flaky. If there is a better way we would gladly try it out.

**Comment by shwarcu** [comment]
@halexiev-hedgeserv as far as I understand, `page.evaluate()` can be used in the following manner:
1. You expose in your app, list of items passed to virtualized/non-virtualized list. For example under object `window.aut.viewXYZ.listItems`
2. In the test, you use page.evaluate to retrieve list of items like such
```js
const items = await page.evaluate(() => window.aut.viewXYZ.listItems);
```
3. You can now run assertions on that list of items, skipping the frontend aspect of test
```js
expect(items).toHaveLength(expectedItemsLength);
```
This would allow you to test list of items, right before they are put into the scrollable list component.

If you really need to test the component itself, in the e2e test, you need to work with actions like scrolling etc. 
Of course this is not a definitive answer to this topic. Someone else may suggest a better approach here.

**Comment by halexiev-hedgeserv** [comment]
@shwarcu Thank you for the explanation. We'll definitely give it a try! One downside of this approach though that could make it a bit cumbersome to implement on a larger scale is that adding properties in the global scope is a bit hacky and would most likely need to be feature flagged so that it does not leak in prod. Anyways, until there's a better approach we'll have a go with this one :)

**Comment by dgozman** [comment]
It seems like the question has been resolved.
